# Explore here

In [1]:
import pandas as pd


# URL del dataset
url = "https://storage.googleapis.com/breathecode/project-files/bank-marketing-campaign-data.csv"

# Cargar el CSV en un DataFrame
df = pd.read_csv(url, sep=";")

# Ver las primeras filas para comprobar que se cargó bien
df.head()


age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [2]:
df.shape

(41188, 21)

In [3]:
df.info

<bound method DataFrame.info of        age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact mo

El problema del banco se traduce en un problema de Machine Learning de clasificación binaria: predecir si un cliente contratará o no un depósito a largo plazo.

Cada fila del dataset representa un intento de contacto (campaña) a un cliente, no al cliente “en abstracto”.

Las variables se agrupan conceptualmente en:

Perfil del cliente (edad, trabajo, estudios, préstamos, etc.).

Información de la campaña (cuándo y cómo se contacta, historial de contactos).

Contexto macroeconómico (indicadores económicos del momento).

La variable objetivo (target) indica el resultado final de la campaña y no debe usarse como entrada del modelo.

Desde este paso ya queda claro que:

No todas las variables serán útiles.

Habrá que vigilar posibles fugas de información (variables que solo se conocen después de la llamada).

El análisis debe mantener siempre el sentido de negocio: decidir a quién llamar antes de hacerlo.

In [4]:
df.head()


age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [5]:
df.tail()


age          job  marital            education default housing loan  \
41183   73      retired  married  professional.course      no     yes   no   
41184   46  blue-collar  married  professional.course      no      no   no   
41185   56      retired  married    university.degree      no     yes   no   
41186   44   technician  married  professional.course      no      no   no   
41187   74      retired  married  professional.course      no     yes   no   

        contact month day_of_week  ...  campaign  pdays  previous  \
41183  cellular   nov         fri  ...         1    999         0   
41184  cellular   nov         fri  ...         1    999         0   
41185  cellular   nov         fri  ...         2    999         0   
41186  cellular   nov         fri  ...         1    999         0   
41187  cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[5 rows x 21 columns]

In [6]:
df.shape


(41188, 21)

In [7]:
df.columns


Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='str')

In [8]:
df.sample(5, random_state=42)


age          job  marital    education  default housing loan  \
32884   57   technician  married  high.school       no      no  yes   
3169    55      unknown  married      unknown  unknown     yes   no   
32206   33  blue-collar  married     basic.9y       no      no   no   
9403    36       admin.  married  high.school       no      no   no   
14020   27    housemaid  married  high.school       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
32884   cellular   may         mon  ...         1    999         1   
3169   telephone   may         thu  ...         2    999         0   
32206   cellular   may         fri  ...         1    999         1   
9403   telephone   jun         fri  ...         4    999         0   
14020   cellular   jul         fri  ...         2    999         0   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
32884      failure         -1.8          92.893          -46.2      1.299   
3169   nonexistent          1.1          93.994          -36.4      4.860   
32206      failure         -1.8          92.893          -46.2      1.313   
9403   nonexistent          1.4          94.465          -41.8      4.967   
14020  nonexistent          1.4          93.918          -42.7      4.963   

       nr.employed   y  
32884       5099.1  no  
3169        5191.0  no  
32206       5099.1  no  
9403        5228.1  no  
14020       5228.1  no  

[5 rows x 21 columns]

In [9]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  str    
 2   marital         41188 non-null  str    
 3   education       41188 non-null  str    
 4   default         41188 non-null  str    
 5   housing         41188 non-null  str    
 6   loan            41188 non-null  str    
 7   contact         41188 non-null  str    
 8   month           41188 non-null  str    
 9   day_of_week     41188 non-null  str    
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  str    
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null  float64
 1

Conclusión paso 3

Revisión de estructura y tipos (interpretación)
📐 Tamaño del dataset

41.188 filas → 41.188 contactos de campaña.

21 columnas → 21 variables.

Esto es un tamaño muy adecuado para un modelo de clasificación.

🧩 Tipos de datos detectados
🔢 Variables numéricas (int64 + float64) → 11

int64 (6):

age

duration

campaign

pdays

previous

float64 (5):

emp.var.rate

cons.price.idx

cons.conf.idx

euribor3m

nr.employed

📌 Estas variables:

son candidatas directas a features numéricas,

pero algunas requerirán revisión por sentido de negocio (ej. duration).

🏷️ Variables categóricas (object) → 11

job

marital

education

default

housing

loan

contact

month

day_of_week

poutcome

y (target)

📌 Importante:

Estas variables no pueden entrar tal cual en un modelo ML.

Más adelante habrá que:

codificarlas,

agrupar categorías,

o eliminarlas si no aportan valor.

✅ Valores no nulos

Todas las columnas tienen 41.188 valores no nulos.

⚠️ Ojo conceptual (lo veremos en el paso 4):

Aunque no haya NaN, pueden existir valores tipo "unknown"
→ no son NaN, pero sí valores faltantes semánticos.

⚠️ Señales importantes detectadas ya en este paso
🔴 duration

Numérica (int64)

Representa duración de la llamada.

📌 Alerta temprana:

Se conoce después de llamar al cliente.

Riesgo alto de data leakage.

Muy predictiva, pero no usable en producción.

(No la tocamos aún, solo la señalamos mentalmente).

🟡 pdays

Días desde último contacto.

Suele usar valores especiales (ej. 999) para “nunca contactado”.

Puede generar valores artificiales.

In [10]:
df.isna().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

In [11]:
df.shape[0] - df.count()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

In [12]:
(df=="unknown").sum()

age                  0
job                330
marital             80
education         1731
default           8597
housing            990
loan               990
contact              0
month                0
day_of_week          0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx       0
cons.conf.idx        0
euribor3m            0
nr.employed          0
y                    0
dtype: int64

In [13]:
(df == "unknown").sum().sort_values(ascending=False)

default           8597
education         1731
housing            990
loan               990
job                330
marital             80
age                  0
contact              0
month                0
day_of_week          0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx       0
cons.conf.idx        0
euribor3m            0
nr.employed          0
y                    0
dtype: int64

In [14]:
unknown_counts = (df == "unknown").sum()
unknown_counts[unknown_counts > 0]

job           330
marital        80
education    1731
default      8597
housing       990
loan          990
dtype: int64

In [15]:
unknown_mask = (df == "unknown")


In [16]:
unknown_per_row = unknown_mask.sum(axis=1)
unknown_per_row

0        0
1        1
2        0
3        0
4        0
        ..
41183    0
41184    0
41185    0
41186    0
41187    0
Length: 41188, dtype: int64

In [17]:
df[unknown_per_row > 1].head()

age          job  marital          education  default  housing     loan  \
7    41  blue-collar  married            unknown  unknown       no       no   
10   41  blue-collar  married            unknown  unknown       no       no   
29   55      unknown  married  university.degree  unknown  unknown  unknown   
35   55      unknown  married           basic.4y  unknown      yes       no   
73   57      unknown  married            unknown  unknown       no       no   

      contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
7   telephone   may         mon  ...         1    999         0  nonexistent   
10  telephone   may         mon  ...         1    999         0  nonexistent   
29  telephone   may         mon  ...         1    999         0  nonexistent   
35  telephone   may         mon  ...         1    999         0  nonexistent   
73  telephone   may         mon  ...         1    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
7           1.1          93.994          -36.4      4.857       5191.0  no  
10          1.1          93.994          -36.4      4.857       5191.0  no  
29          1.1          93.994          -36.4      4.857       5191.0  no  
35          1.1          93.994          -36.4      4.857       5191.0  no  
73          1.1          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [18]:
unknown_per_row.value_counts().sort_index()

0    30488
1     9034
2     1338
3      306
4       20
5        2
Name: count, dtype: int64

In [19]:
unknown_per_row.max()

np.int64(5)

In [20]:
row_idx = unknown_per_row[unknown_per_row > 4].index[0]

df.loc[row_idx]

age                        45
job                   unknown
marital               married
education             unknown
default               unknown
housing               unknown
loan                  unknown
contact             telephone
month                     may
day_of_week               wed
duration                  179
campaign                    1
pdays                     999
previous                    0
poutcome          nonexistent
emp.var.rate              1.1
cons.price.idx         93.994
cons.conf.idx           -36.4
euribor3m               4.859
nr.employed            5191.0
y                          no
Name: 2723, dtype: object

In [21]:
df.loc[row_idx, unknown_mask.loc[row_idx]]

job          unknown
education    unknown
default      unknown
housing      unknown
loan         unknown
Name: 2723, dtype: str

Conclusión técnica del PASO 4

No hay valores nulos (NaN), pero:

Sí hay valores faltantes codificados como texto.

Estos valores:

no pueden ignorarse,

afectan directamente al modelo.

📌 En Machine Learning:

Tratar mal los "unknown" suele degradar el rendimiento más que los NaN.

5️⃣ Decisiones (solo conceptuales, aún no ejecutamos)

En este paso no transformamos nada, pero dejamos claro:

Qué variables tienen problemas de completitud.

Cuáles requieren decisión explícita más adelante.

Que "unknown" ≠ categoría real.

In [22]:
df.duplicated().sum()


np.int64(12)

In [23]:
df = df.drop_duplicates()


CONCLUSIÓN PASO 5

Se detectaron 12 duplicados exactos.

Son duplicaciones técnicas, no casos reales distintos.

La decisión correcta es eliminarlos.

El dataset queda más consistente para el análisis posterior.

In [24]:
df.info()

<class 'pandas.DataFrame'>
Index: 41176 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41176 non-null  int64  
 1   job             41176 non-null  str    
 2   marital         41176 non-null  str    
 3   education       41176 non-null  str    
 4   default         41176 non-null  str    
 5   housing         41176 non-null  str    
 6   loan            41176 non-null  str    
 7   contact         41176 non-null  str    
 8   month           41176 non-null  str    
 9   day_of_week     41176 non-null  str    
 10  duration        41176 non-null  int64  
 11  campaign        41176 non-null  int64  
 12  pdays           41176 non-null  int64  
 13  previous        41176 non-null  int64  
 14  poutcome        41176 non-null  str    
 15  emp.var.rate    41176 non-null  float64
 16  cons.price.idx  41176 non-null  float64
 17  cons.conf.idx   41176 non-null  float64
 18  eu

In [25]:
df.describe()

age      duration      campaign         pdays      previous  \
count  41176.00000  41176.000000  41176.000000  41176.000000  41176.000000   
mean      40.02380    258.315815      2.567879    962.464810      0.173013   
std       10.42068    259.305321      2.770318    186.937102      0.494964   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41176.000000    41176.000000   41176.000000  41176.000000  41176.000000  
mean       0.081922       93.575720     -40.502863      3.621293   5167.034870  
std        1.570883        0.578839       4.627860      1.734437     72.251364  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

* age

Rango: 17–98

Mediana: 38 | Media: ~40
Lectura: rango plausible; ligera asimetría positiva (media > mediana).
Conclusión parcial: estable, sin outliers evidentes.

* duration ⚠️

Rango: 0–4918 (≈ 82 min)

Mediana: 180 | Media: 258
Lectura: distribución muy sesgada; valores extremos claros.
Alerta de negocio: se conoce después de la llamada → riesgo de data leakage.
Conclusión parcial: muy informativa pero problemática.

* campaign

Rango: 1–56

Mediana: 2 | 75%: 3
Lectura: la mayoría recibe pocas llamadas; hay casos extremos.
Conclusión parcial: outliers conductuales (no necesariamente errores).

* pdays ⚠️

Cuartiles (25%, 50%, 75%): 999
Lectura clave: 999 = “nunca contactado”.
Conclusión parcial: variable codificada, no continua real.

* previous

Rango: 0–7 | Mediana: 0
Lectura: mayoría sin contactos previos.
Conclusión parcial: estable.

* Variables macroeconómicas

- emp.var.rate: -3.4 a 1.4

- cons.price.idx: ~92.2–94.8

- cons.conf.idx: valores negativos esperables

- euribor3m: ~0.63–5.04

- nr.employed: ~4963–5228

Lectura conjunta:

Sin outliers evidentes.

Escalas muy distintas entre variables.

| Variable   | Señal                       |
| ---------- | --------------------------- |
| `duration` | Sesgo extremo + leakage     |
| `pdays`    | Codificación especial (999) |
| `campaign` | Outliers conductuales       |
| `age`      | Normal                      |
| Macro vars | Estables, distintas escalas |


Conclusión del PASO 6

Ya conocemos rangos y dispersión de las variables numéricas.

Hemos marcado variables críticas para revisar después (duration, pdays).

El dataset refleja comportamiento real, no ruido arbitrario.

Quedan pendientes decisiones técnicas y de negocio para pasos posteriores.

In [26]:
df.select_dtypes(include="object").nunique().sort_values(ascending=False)


/tmp/ipykernel_1085/607924466.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.select_dtypes(include="object").nunique().sort_values(ascending=False)


job            12
month          10
education       8
day_of_week     5
marital         4
default         3
housing         3
poutcome        3
loan            3
contact         2
y               2
dtype: int64

Resultado observado

| Variable      | Nº categorías | Lectura            |
| ------------- | ------------- | ------------------ |
| `job`         | 12            | Alta diversidad    |
| `month`       | 10            | Estacional         |
| `education`   | 8             | Ordinal            |
| `day_of_week` | 5             | Cíclica            |
| `marital`     | 4             | Simple             |
| `default`     | 3             | Incluye `unknown`  |
| `housing`     | 3             | Incluye `unknown`  |
| `poutcome`    | 3             | Muy desequilibrada |
| `loan`        | 3             | Incluye `unknown`  |
| `contact`     | 2             | Muy limpia         |
| `y`           | 2             | Target binario     |


🔹 Variables con muchas categorías

job (12)

Riesgo: categorías poco frecuentes.

Puede fragmentar la información.

🔹 Variables temporales

month (10)

day_of_week (5)

📌 No son ordinales reales, son cíclicas
(Enero no está “lejos” de Diciembre).

🔹 Variable ordinal clara

education (8)
Tiene orden lógico, aunque venga como texto.

🔹 Variables binarias / casi binarias

contact, loan, housing, default, poutcome

📌 Simples de tratar, pero:

default, loan, housing incluyen "unknown"
→ ya identificado como valor faltante semántico.

🔹 Target y

2 valores (yes, no)

Correcto para clasificación binaria.

Conclusión del PASO 7 (solo con esta salida)

- Ya conocemos la complejidad categórica de cada variable.

- Sabemos:

    - cuáles son simples,

    - cuáles  son complejas,

    - cuáles tienen orden implícito,

    - cuáles contienen valores problemáticos (unknown).

Este paso no implica decisiones, solo entender la estructura categórica.

1️⃣ ¿Qué variables de tiempo hay en el dataset?

En este dataset NO hay fechas completas (tipo YYYY-MM-DD).
Las variables relacionadas con tiempo son:

month

day_of_week

(parcialmente) pdays → ya analizada como numérica especial

2️⃣ Análisis de month
Qué es

Mes en el que se realizó la llamada (jan, feb, …).

Tipo real

Variable categórica temporal.

No ordinal en sentido numérico.

Cíclica:

enero está “cerca” de diciembre.

📌 Importante:

Aunque tenga orden natural, no debe tratarse como 1–12 sin cuidado.

🧠 Negocio:

Las campañas funcionan distinto según la época del año.

Captura estacionalidad.

3️⃣ Análisis de day_of_week
Qué es

Día de la semana de la llamada (mon, tue, …).

Tipo real

Variable categórica temporal.

Cíclica:

lunes y viernes no son extremos opuestos.

📌 Igual que month:

No es ordinal para ML de forma directa.

Representa patrones operativos del call center.

4️⃣ Qué NO hay (y por qué importa)

No hay:

fecha exacta,

hora del día,

timestamp.

📌 Implicación:

No podemos analizar:

tendencias temporales continuas,

series temporales.

El enfoque sigue siendo tabular, no temporal.

5️⃣ Relación con pasos anteriores

month y day_of_week:

ya sabemos que tienen pocas categorías (PASO 7),

no tienen valores "unknown" (PASO 4),

no tienen outliers numéricos (PASO 6).

Son variables:

limpias,

estables,

pero requieren tratamiento especial por su naturaleza cíclica.

Conclusión del PASO 8

El dataset no contiene fechas completas, solo categorías temporales.

month y day_of_week:

representan estacionalidad,

son cíclicas, no ordinales simples.

Son potencialmente útiles para el modelo,

pero deberán tratarse con cuidado más adelante.

En este paso solo se identifican y entienden, no se transforman.

Conclusión del PASO 9

El dataset no contiene texto libre.

Todas las variables tipo texto son categóricas controladas.

No se requieren técnicas de NLP.

Este paso queda cerrado sin acciones.

In [27]:
df["y"].value_counts()


y
no     36537
yes     4639
Name: count, dtype: int64

In [28]:
df["y"].value_counts(normalize=True)


y
no     0.887337
yes    0.112663
Name: proportion, dtype: float64

In [29]:
df.groupby("job")["y"].value_counts(normalize=True).unstack()


y                    no       yes
job                              
admin.         0.870333  0.129667
blue-collar    0.931049  0.068951
entrepreneur   0.914835  0.085165
housemaid      0.900000  0.100000
management     0.887825  0.112175
retired        0.747381  0.252619
self-employed  0.895144  0.104856
services       0.918578  0.081422
student        0.685714  0.314286
technician     0.891675  0.108325
unemployed     0.857988  0.142012
unknown        0.887879  0.112121

In [30]:
df.groupby("education")["y"].value_counts(normalize=True).unstack()


y                          no       yes
education                              
basic.4y             0.897510  0.102490
basic.6y             0.917940  0.082060
basic.9y             0.921754  0.078246
high.school          0.891611  0.108389
illiterate           0.777778  0.222222
professional.course  0.886450  0.113550
university.degree    0.862792  0.137208
unknown              0.854913  0.145087

In [31]:
df.groupby("poutcome")["y"].value_counts(normalize=True).unstack()


y                  no       yes
poutcome                       
failure      0.857714  0.142286
nonexistent  0.911676  0.088324
success      0.348871  0.651129

In [32]:
df.groupby("y")["age"].describe()


count       mean        std   min   25%   50%   75%   max
y                                                               
no   36537.0  39.910994   9.897176  17.0  32.0  38.0  47.0  95.0
yes   4639.0  40.912266  13.838838  17.0  31.0  37.0  50.0  98.0

In [33]:
df.groupby("y")["campaign"].describe()


count      mean       std  min  25%  50%  75%   max
y                                                         
no   36537.0  2.633385  2.873768  1.0  1.0  2.0  3.0  56.0
yes   4639.0  2.051951  1.666353  1.0  1.0  2.0  2.0  23.0

Conclusión del PASO 10

Ya conocemos:

el balance del target,

qué variables parecen más relacionadas con y,

qué patrones de negocio aparecen en los datos.

Hemos identificado:

señales predictivas potenciales,

posibles riesgos de fuga de información.

El análisis sigue siendo descriptivo, no predictivo.

In [34]:
num_df = df.select_dtypes(include=["int64", "float64"])
corr = num_df.corr()
corr


age  duration  campaign     pdays  previous  \
age             1.000000 -0.000808  0.004622 -0.034381  0.024379   
duration       -0.000808  1.000000 -0.071765 -0.047556  0.020600   
campaign        0.004622 -0.071765  1.000000  0.052606 -0.079182   
pdays          -0.034381 -0.047556  0.052606  1.000000 -0.587508   
previous        0.024379  0.020600 -0.079182 -0.587508  1.000000   
emp.var.rate   -0.000242 -0.027941  0.150786  0.271063 -0.420587   
cons.price.idx  0.001009  0.005303  0.127826  0.078920 -0.203197   
cons.conf.idx   0.129075 -0.008126 -0.013657 -0.091374 -0.050929   
euribor3m       0.010852 -0.032861  0.135169  0.296946 -0.454571   
nr.employed    -0.017607 -0.044672  0.144129  0.372659 -0.501411   

                emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
age                -0.000242        0.001009       0.129075   0.010852   
duration           -0.027941        0.005303      -0.008126  -0.032861   
campaign            0.150786        0.127826      -0.013657   0.135169   
pdays               0.271063        0.078920      -0.091374   0.296946   
previous           -0.420587       -0.203197      -0.050929  -0.454571   
emp.var.rate        1.000000        0.775293       0.196257   0.972244   
cons.price.idx      0.775293        1.000000       0.059170   0.688180   
cons.conf.idx       0.196257        0.059170       1.000000   0.277864   
euribor3m           0.972244        0.688180       0.277864   1.000000   
nr.employed         0.906949        0.521945       0.100679   0.945146   

                nr.employed  
age               -0.017607  
duration          -0.044672  
campaign           0.144129  
pdays              0.372659  
previous          -0.501411  
emp.var.rate       0.906949  
cons.price.idx     0.521945  
cons.conf.idx      0.100679  
euribor3m          0.945146  
nr.employed        1.000000

🔹 Bloque 1: Variables del cliente / campaña
age

Correlaciones casi nulas con todas las demás.

📌 Lectura:

Edad aporta información independiente.

No hay multicolinealidad asociada.

campaign

Correlaciones bajas con casi todo.

Máx ≈ 0.15 con variables macro.

📌 Lectura:

No es redundante.

Relación débil con el contexto económico.

duration

Correlaciones muy bajas con todas.

📌 Lectura técnica:

No está linealmente correlacionada con otras numéricas.

⚠️ Nota conceptual (sin decidir):

Su problema no es multicolinealidad, sino data leakage (visto antes).

🔹 Bloque 2: Variables de historial (pdays, previous)
pdays

Correlación negativa moderada con previous (~ -0.59).

📌 Lectura:

Tiene sentido:

si pdays es alto (999 = nunca),

entonces previous suele ser 0.

⚠️ Importante:

Esta correlación está artificialmente inducida por el valor 999.

previous

Correlaciones negativas con:

emp.var.rate

euribor3m

nr.employed

📌 Lectura:

El historial de campañas se da en contextos económicos concretos.

No es un problema técnico, sino contextual.

Bloque 3: Variables macroeconómicas ⚠️

Aquí está el punto clave del PASO 11.

Correlaciones muy altas detectadas:

Variables	Correlación
- emp.var.rate ↔ euribor3m	0.97
- euribor3m ↔ nr.employed	0.95
- emp.var.rate ↔ nr.employed	0.91
- emp.var.rate ↔ cons.price.idx	0.77
- euribor3m ↔ cons.price.idx	0.69

📌 Esto es multicolinealidad clara.

Conclusión del PASO 11

No hay multicolinealidad problemática entre:

variables del cliente,

variables de campaña.

Sí hay multicolinealidad fuerte entre variables macroeconómicas.

Esta multicolinealidad:

es esperable,

debe tenerse en cuenta en decisiones futuras,

no se corrige todavía.

Análisis variable por variable (según tus estadísticas)
🔹 age

Rango: 17 – 98

IQR aproximado: 47 − 32 = 15

Límites:

Inferior ≈ 32 − 22.5 ≈ 9.5

Superior ≈ 47 + 22.5 ≈ 69.5

📌 Valores > ~70 podrían considerarse outliers técnicos.

Pero:

Edad 80–90 es posible.

No parece error.

✔️ Conclusión: outliers reales, no errores.

🔹 duration ⚠️

Max: 4918

Mediana: 180

Muy asimétrica.

Aquí sí hay outliers extremos.

Pero:

No son errores.

Son llamadas largas reales.

🚨 Recordatorio:

El problema no es el outlier.

El problema es data leakage.

🔹 campaign

Q3 = 3

Max = 56

IQR ≈ 3 − 1 = 2
Límite superior ≈ 3 + 3 = 6

Todo lo >6 es outlier técnico.

Pero:

56 llamadas no es error.

Es comportamiento extremo del call center.

✔️ Outlier conductual, no error.

🔹 pdays ⚠️

75% = 999

Max = 999

Esto no es un outlier.

Es:

una codificación artificial.

📌 No debe tratarse como continuo real.

🔹 previous

Max = 7

Q3 = 0

Outliers técnicos posibles (>1 o 2).

Pero:

No son errores.

Son clientes muy insistidos.

🔹 Variables macroeconómicas

No presentan extremos anómalos.

Distribuciones estables.

Sin valores sospechosos.

✔️ No problemáticas en términos de outliers.

| Variable   | Tipo de problema            |
| ---------- | --------------------------- |
| `age`      | Outliers reales, aceptables |
| `duration` | Extremos reales + leakage   |
| `campaign` | Outliers conductuales       |
| `pdays`    | Valor artificial (999)      |
| `previous` | Valores extremos leves      |
| Macro vars | Sin problemas               |


Conclusión del PASO 12

No se detectan errores de datos.

Sí existen valores extremos reales.

pdays contiene codificación artificial.

duration es problemática por leakage, no por outlier.

El dataset refleja comportamiento real del negocio.

Variables problemáticas por sentido de negocio
🔴 duration

Se conoce después de la llamada.

Es consecuencia del interés del cliente.

Introduce data leakage.

📌 Desde negocio:

No podemos saber cuánto durará una llamada antes de hacerla.

🟡 pdays

999 significa “nunca contactado”.

No es un valor continuo real.

Mezcla dos significados:

días reales,

categoría “nunca”.

📌 Problema:

Si no se transforma, el modelo interpretará 999 como un número grande real.

🟡 default

~21% "unknown".

El banco no sabe si el cliente tiene impagos.

📌 Problema:

Variable poco fiable.

Puede introducir ruido.

2️⃣ Variables coherentes con negocio
🟢 Perfil del cliente

age

job

marital

education

housing

loan

✔️ Información disponible antes de llamar.

🟢 Historial de campañas

campaign

previous

poutcome

✔️ Información real del CRM.

🟢 Variables macroeconómicas

euribor3m

emp.var.rate

nr.employed

etc.

✔️ Información pública disponible en el momento de la campaña.

3️⃣ Calidad estructural del dataset
✔️ Puntos positivos

Sin valores nulos reales.

Solo 12 duplicados (ya tratados).

Dataset grande (41k filas).

Variables coherentes con el problema.

⚠️ Puntos a vigilar

"unknown" como missing semántico.

Codificación artificial (pdays = 999).

Multicolinealidad macroeconómica.

Variable con leakage (duration).

4️⃣ Pregunta clave de negocio

El modelo debe responder:

¿A quién llamo antes de llamar?

Por tanto:

Solo podemos usar variables disponibles ex ante.

No podemos usar consecuencias de la llamada.

5️⃣ Evaluación general de calidad

El dataset es:

Realista.

Consistente.

Representa un problema de negocio auténtico.

Contiene variables muy útiles.

Pero necesita limpieza y decisiones estratégicas.

✅ Conclusión del PASO 13

La mayoría de variables tienen sentido para el negocio.

duration introduce leakage.

pdays necesita transformación.

default tiene baja calidad informativa.

Existe multicolinealidad estructural en variables macro.

El dataset es adecuado para construir un modelo robusto si se toman buenas decisiones.

# Conclusiones y decisiones finales del EDA

Objetivo

- Qué variables se eliminan.

- Qué variables se transforman.

- Qué variables se mantienen.

- Qué problemas debemos corregir antes del modelo.

1️⃣ Variables a ELIMINAR (Drop)
🔴 duration

Motivo:

Es data leakage.

Solo se conoce después de la llamada.

Inflaría artificialmente el rendimiento del modelo.

2️⃣ Variables que requieren TRANSFORMACIÓN
🟡 pdays

Problema:

999 significa “nunca contactado”.

Decisión conceptual:

Convertir en:

Variable binaria: previous_contacted (sí/no)

O separar:

never_contacted

days_since_last_contact real

No dejarlo como continuo puro.

🟡 Variables con "unknown"

Afectadas:

default

education

housing

loan

job

marital

Decisión conceptual:

Mantener "unknown" como categoría válida
O

Imputar estratégicamente (según modelo).

⚠️ Caso especial: default

Muy alto porcentaje de "unknown".

Candidata a eliminar si aporta poco valor predictivo.

3️⃣ Variables con Multicolinealidad

Bloque macroeconómico:

emp.var.rate

euribor3m

nr.employed

cons.price.idx

Decisión conceptual:

Mantener todas inicialmente.

Evaluar después:

Regularización (L2),

Selección de variables,

Eliminación de redundantes.

No eliminar aún sin medir impacto.

4️⃣ Variables que se mantienen sin cambios

✔️ age
✔️ campaign
✔️ previous
✔️ poutcome
✔️ contact
✔️ month
✔️ day_of_week
✔️ Variables macro (con cautela)

5️⃣ Variables categóricas → requerirán codificación

Todas las object deberán:

One-Hot Encoding (nominales)

Codificación ordinal (education)

Tratamiento especial para variables cíclicas (month, day_of_week)

Pero eso pertenece al preprocesamiento, no al EDA.

6️⃣ Outliers

Decisión:

No eliminar outliers reales.

No se detectaron errores de medición.

campaign extremos reflejan comportamiento real.

7️⃣ Estado final del dataset tras EDA

El dataset queda conceptualmente:

Sin duplicados.

Sin leakage.

Con variables estructuralmente entendidas.

Con problemas identificados y plan de acción claro.

In [35]:
df = df.rename(columns={
    "age": "edad_cliente",
    "job": "ocupacion",
    "marital": "estado_civil",
    "education": "nivel_educacion",
    "default": "credito_actual",
    "housing": "prestamo_hipotecario",
    "loan": "prestamo_personal",
    "contact": "tipo_contacto",
    "month": "ultimo_mes_contacto",
    "day_of_week": "ultimo_dia_contacto",
    "duration": "duracion_contacto_segundos",
    "campaign": "num_contactos_campana_actual",
    "pdays": "num_dias_desde_ultima_campana",
    "previous": "num_contactos_campana_anterior",
    "poutcome": "resultado_campana_anterior",
    "emp.var.rate": "tasa_variacion_empleo",
    "cons.price.idx": "indice_precios_consumidor",
    "cons.conf.idx": "indice_confianza_consumidor",
    "euribor3m": "tasa_euribor",
    "nr.employed": "num_empleados",
    "y": "contrata_deposito_largo_plazo"
})


In [36]:
df.columns


Index(['edad_cliente', 'ocupacion', 'estado_civil', 'nivel_educacion',
       'credito_actual', 'prestamo_hipotecario', 'prestamo_personal',
       'tipo_contacto', 'ultimo_mes_contacto', 'ultimo_dia_contacto',
       'duracion_contacto_segundos', 'num_contactos_campana_actual',
       'num_dias_desde_ultima_campana', 'num_contactos_campana_anterior',
       'resultado_campana_anterior', 'tasa_variacion_empleo',
       'indice_precios_consumidor', 'indice_confianza_consumidor',
       'tasa_euribor', 'num_empleados', 'contrata_deposito_largo_plazo'],
      dtype='str')

In [37]:
df = df.drop(columns=["duracion_contacto_segundos"])


In [38]:
"duracion_contacto_segundos" in df.columns


False

Eliminamos duracion_contacto_segundos porque es data leakage: su valor solo se conoce después de realizar la llamada, y nuestro modelo pretende decidir a quién llamar antes de hacerlo.

Usarla permitiría al modelo aprender una consecuencia del resultado (las llamadas exitosas suelen durar más), inflando artificialmente el rendimiento y generando un modelo que no sería válido en producción.

In [39]:
df["num_dias_desde_ultima_campana"].value_counts().sort_index()


num_dias_desde_ultima_campana
0         15
1         26
2         61
3        439
4        118
5         46
6        412
7         60
8         18
9         64
10        52
11        28
12        58
13        36
14        20
15        24
16        11
17         8
18         7
19         3
20         1
21         2
22         3
25         1
26         1
27         1
999    39661
Name: count, dtype: int64

In [40]:
import numpy as np

# 1) Flag: nunca contactado antes (True/False)
df["nunca_contactado_antes"] = df["num_dias_desde_ultima_campana"].eq(999)

# 2) Reemplazar 999 por NaN para que la variable sea "días reales" cuando aplica
df["num_dias_desde_ultima_campana"] = df["num_dias_desde_ultima_campana"].replace(999, np.nan)


In [41]:
df[["num_dias_desde_ultima_campana", "nunca_contactado_antes"]].head(10)

num_dias_desde_ultima_campana  nunca_contactado_antes
0                            NaN                    True
1                            NaN                    True
2                            NaN                    True
3                            NaN                    True
4                            NaN                    True
5                            NaN                    True
6                            NaN                    True
7                            NaN                    True
8                            NaN                    True
9                            NaN                    True

In [42]:

df["num_dias_desde_ultima_campana"].isna().sum()
df["nunca_contactado_antes"].value_counts()


nunca_contactado_antes
True     39661
False     1515
Name: count, dtype: int64

# Transformación de num_dias_desde_ultima_campana

El análisis mostró que aproximadamente el 96% de los registros corresponden a clientes nunca contactados previamente, lo que evidencia que el valor 999 no representa un número real de días, sino una codificación artificial de ausencia de historial.

Mantener 999 como valor numérico habría distorsionado la interpretación del modelo, al tratar una condición categórica (“nunca contactado”) como si fuera una magnitud continua extrema.

La transformación aplicada:

Separó correctamente la dimensión conceptual en:

una variable binaria (nunca_contactado_antes),

y una variable numérica válida solo cuando existe historial real.

Preserva la semántica del negocio.

Reduce el riesgo de sesgo estructural.

Mejora la calidad de la información que recibirá el modelo.

En términos de ingeniería de datos, este paso convierte una variable mal codificada en una representación coherente, interpretable y adecuada para modelado predictivo.

Objetivo

Tratar correctamente los valores "unknown" en:

ocupacion

estado_civil

nivel_educacion

credito_actual

prestamo_hipotecario

prestamo_personal

1️⃣ Pregunta clave antes de decidir

¿"unknown" significa lo mismo en todas las columnas?

No.

Puede significar:

El cliente no quiso responder.

El banco no tiene el dato.

El dato no fue registrado.

El valor no aplica.

Por tanto, debemos tratarlos estratégicamente.

2️⃣ Clasificación por tipo de variable
🟡 Grupo 1 — Variables demográficas

ocupacion

estado_civil

nivel_educacion

Aquí "unknown" significa:

No tenemos la información del perfil.

📌 Recomendación profesional:
Mantener "unknown" como categoría válida.

¿Por qué?

Es información en sí misma.

Puede correlacionar con menor probabilidad de contratación.

No distorsiona el modelo.

🔴 Grupo 2 — credito_actual

Aquí "unknown" representa casi 21%.

Problema:

Alta incertidumbre.

Puede introducir ruido.

Opciones:

Mantener "unknown" como categoría.

Convertir "unknown" en NaN.

Eliminar la variable si aporta poco.

📌 Por ahora:
Mantener como categoría y evaluar más adelante con métricas.

🟡 Grupo 3 — prestamo_hipotecario y prestamo_personal

Tienen "unknown" moderado.

Recomendación:
Mantener como categoría.
Puede reflejar perfil financiero incompleto.

3️⃣ ¿Debemos convertir "unknown" en NaN?

No necesariamente.

Diferencia clave:

NaN	"unknown"
Dato faltante real	Información categórica explícita

En este dataset:

"unknown" es una categoría informativa, no un error.

Por tanto:

👉 Mantenerlo como categoría es coherente.

4️⃣ Conclusión profesional del tratamiento de "unknown"

No se trata de valores faltantes técnicos.

Son categorías explícitas del dataset.

Mantenerlos preserva información estructural.

No requieren imputación en esta fase.

Se tratarán en la codificación (One-Hot u ordinal).

# Estado tras este paso

duration eliminado.

pdays transformado correctamente.

"unknown" evaluado y mantenido estratégicamente.

Dataset listo para fase de preprocesamiento formal.

In [43]:
import sklearn
import matplotlib.pyplot as plt
import seaborn as sns

In [44]:
df.head()

edad_cliente  ocupacion estado_civil nivel_educacion credito_actual  \
0            56  housemaid      married        basic.4y             no   
1            57   services      married     high.school        unknown   
2            37   services      married     high.school             no   
3            40     admin.      married        basic.6y             no   
4            56   services      married     high.school             no   

  prestamo_hipotecario prestamo_personal tipo_contacto ultimo_mes_contacto  \
0                   no                no     telephone                 may   
1                   no                no     telephone                 may   
2                  yes                no     telephone                 may   
3                   no                no     telephone                 may   
4                   no               yes     telephone                 may   

  ultimo_dia_contacto  ...  num_dias_desde_ultima_campana  \
0                 mon  ...                            NaN   
1                 mon  ...                            NaN   
2                 mon  ...                            NaN   
3                 mon  ...                            NaN   
4                 mon  ...                            NaN   

   num_contactos_campana_anterior  resultado_campana_anterior  \
0                               0                 nonexistent   
1                               0                 nonexistent   
2                               0                 nonexistent   
3                               0                 nonexistent   
4                               0                 nonexistent   

  tasa_variacion_empleo  indice_precios_consumidor  \
0                   1.1                     93.994   
1                   1.1                     93.994   
2                   1.1                     93.994   
3                   1.1                     93.994   
4                   1.1                     93.994   

   indice_confianza_consumidor  tasa_euribor  num_empleados  \
0                        -36.4         4.857         5191.0   
1                        -36.4         4.857         5191.0   
2                        -36.4         4.857         5191.0   
3                        -36.4         4.857         5191.0   
4                        -36.4         4.857         5191.0   

   contrata_deposito_largo_plazo nunca_contactado_antes  
0                             no                   True  
1                             no                   True  
2                             no                   True  
3                             no                   True  
4                             no                   True  

[5 rows x 21 columns]

In [45]:
df[["num_dias_desde_ultima_campana","resultado_campana_anterior"]]

num_dias_desde_ultima_campana resultado_campana_anterior
0                                NaN                nonexistent
1                                NaN                nonexistent
2                                NaN                nonexistent
3                                NaN                nonexistent
4                                NaN                nonexistent
...                              ...                        ...
41183                            NaN                nonexistent
41184                            NaN                nonexistent
41185                            NaN                nonexistent
41186                            NaN                nonexistent
41187                            NaN                    failure

[41176 rows x 2 columns]

In [46]:
# Modelo A: incluye num_dias_desde_ultima_campana
df_A = df.copy()

# Modelo B: elimina num_dias_desde_ultima_campana
df_B = df.drop(columns=["num_dias_desde_ultima_campana"]).copy()


In [47]:
df_A.head()

edad_cliente  ocupacion estado_civil nivel_educacion credito_actual  \
0            56  housemaid      married        basic.4y             no   
1            57   services      married     high.school        unknown   
2            37   services      married     high.school             no   
3            40     admin.      married        basic.6y             no   
4            56   services      married     high.school             no   

  prestamo_hipotecario prestamo_personal tipo_contacto ultimo_mes_contacto  \
0                   no                no     telephone                 may   
1                   no                no     telephone                 may   
2                  yes                no     telephone                 may   
3                   no                no     telephone                 may   
4                   no               yes     telephone                 may   

  ultimo_dia_contacto  ...  num_dias_desde_ultima_campana  \
0                 mon  ...                            NaN   
1                 mon  ...                            NaN   
2                 mon  ...                            NaN   
3                 mon  ...                            NaN   
4                 mon  ...                            NaN   

   num_contactos_campana_anterior  resultado_campana_anterior  \
0                               0                 nonexistent   
1                               0                 nonexistent   
2                               0                 nonexistent   
3                               0                 nonexistent   
4                               0                 nonexistent   

  tasa_variacion_empleo  indice_precios_consumidor  \
0                   1.1                     93.994   
1                   1.1                     93.994   
2                   1.1                     93.994   
3                   1.1                     93.994   
4                   1.1                     93.994   

   indice_confianza_consumidor  tasa_euribor  num_empleados  \
0                        -36.4         4.857         5191.0   
1                        -36.4         4.857         5191.0   
2                        -36.4         4.857         5191.0   
3                        -36.4         4.857         5191.0   
4                        -36.4         4.857         5191.0   

   contrata_deposito_largo_plazo nunca_contactado_antes  
0                             no                   True  
1                             no                   True  
2                             no                   True  
3                             no                   True  
4                             no                   True  

[5 rows x 21 columns]

In [48]:
df_B.head()

edad_cliente  ocupacion estado_civil nivel_educacion credito_actual  \
0            56  housemaid      married        basic.4y             no   
1            57   services      married     high.school        unknown   
2            37   services      married     high.school             no   
3            40     admin.      married        basic.6y             no   
4            56   services      married     high.school             no   

  prestamo_hipotecario prestamo_personal tipo_contacto ultimo_mes_contacto  \
0                   no                no     telephone                 may   
1                   no                no     telephone                 may   
2                  yes                no     telephone                 may   
3                   no                no     telephone                 may   
4                   no               yes     telephone                 may   

  ultimo_dia_contacto  num_contactos_campana_actual  \
0                 mon                             1   
1                 mon                             1   
2                 mon                             1   
3                 mon                             1   
4                 mon                             1   

   num_contactos_campana_anterior resultado_campana_anterior  \
0                               0                nonexistent   
1                               0                nonexistent   
2                               0                nonexistent   
3                               0                nonexistent   
4                               0                nonexistent   

   tasa_variacion_empleo  indice_precios_consumidor  \
0                    1.1                     93.994   
1                    1.1                     93.994   
2                    1.1                     93.994   
3                    1.1                     93.994   
4                    1.1                     93.994   

   indice_confianza_consumidor  tasa_euribor  num_empleados  \
0                        -36.4         4.857         5191.0   
1                        -36.4         4.857         5191.0   
2                        -36.4         4.857         5191.0   
3                        -36.4         4.857         5191.0   
4                        -36.4         4.857         5191.0   

  contrata_deposito_largo_plazo  nunca_contactado_antes  
0                            no                    True  
1                            no                    True  
2                            no                    True  
3                            no                    True  
4                            no                    True

# Machine Learning

Separamos X e Y para df_A

In [49]:
# Modelo A (con num_dias_desde_ultima_campana)

X_A = df_A.drop(columns=["contrata_deposito_largo_plazo"])
y_A = df_A["contrata_deposito_largo_plazo"]


Separamos X e y para df_B

In [50]:
# Modelo B (sin num_dias_desde_ultima_campana)

X_B = df_B.drop(columns=["contrata_deposito_largo_plazo"])
y_B = df_B["contrata_deposito_largo_plazo"]


In [51]:
print("Modelo A:", X_A.shape, y_A.shape)
print("Modelo B:", X_B.shape, y_B.shape)

y_A.value_counts()


Modelo A: (41176, 20) (41176,)
Modelo B: (41176, 19) (41176,)


contrata_deposito_largo_plazo
no     36537
yes     4639
Name: count, dtype: int64

# EDA para máquinas


Modelo A

In [52]:
col_numericas_A = X_A.select_dtypes(include=['number']).columns.tolist()
col_categoricas_A = X_A.select_dtypes(include=['object']).columns.tolist()

print("Columnas numéricas (A):", col_numericas_A)
print("Columnas categóricas (A):", col_categoricas_A)


Columnas numéricas (A): ['edad_cliente', 'num_contactos_campana_actual', 'num_dias_desde_ultima_campana', 'num_contactos_campana_anterior', 'tasa_variacion_empleo', 'indice_precios_consumidor', 'indice_confianza_consumidor', 'tasa_euribor', 'num_empleados']
Columnas categóricas (A): ['ocupacion', 'estado_civil', 'nivel_educacion', 'credito_actual', 'prestamo_hipotecario', 'prestamo_personal', 'tipo_contacto', 'ultimo_mes_contacto', 'ultimo_dia_contacto', 'resultado_campana_anterior']


/tmp/ipykernel_1085/515039313.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  col_categoricas_A = X_A.select_dtypes(include=['object']).columns.tolist()


Modelo B

In [53]:
col_numericas_B = X_B.select_dtypes(include=['number']).columns.tolist()
col_categoricas_B = X_B.select_dtypes(include=['object']).columns.tolist()

print("Columnas numéricas (B):", col_numericas_B)
print("Columnas categóricas (B):", col_categoricas_B)


Columnas numéricas (B): ['edad_cliente', 'num_contactos_campana_actual', 'num_contactos_campana_anterior', 'tasa_variacion_empleo', 'indice_precios_consumidor', 'indice_confianza_consumidor', 'tasa_euribor', 'num_empleados']
Columnas categóricas (B): ['ocupacion', 'estado_civil', 'nivel_educacion', 'credito_actual', 'prestamo_hipotecario', 'prestamo_personal', 'tipo_contacto', 'ultimo_mes_contacto', 'ultimo_dia_contacto', 'resultado_campana_anterior']


/tmp/ipykernel_1085/738895664.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  col_categoricas_B = X_B.select_dtypes(include=['object']).columns.tolist()


In [54]:
X_A.dtypes["nunca_contactado_antes"]


dtype('bool')

In [55]:
X_A["nunca_contactado_antes"] = X_A["nunca_contactado_antes"].astype(int)
X_B["nunca_contactado_antes"] = X_B["nunca_contactado_antes"].astype(int)


In [56]:
X_A.dtypes["nunca_contactado_antes"]
X_B.dtypes["nunca_contactado_antes"]


dtype('int64')

In [57]:
col_numericas_A = X_A.select_dtypes(include=['number']).columns.tolist()
col_numericas_B = X_B.select_dtypes(include=['number']).columns.tolist()

print("Numéricas A:", col_numericas_A)
print("Numéricas B:", col_numericas_B)


Numéricas A: ['edad_cliente', 'num_contactos_campana_actual', 'num_dias_desde_ultima_campana', 'num_contactos_campana_anterior', 'tasa_variacion_empleo', 'indice_precios_consumidor', 'indice_confianza_consumidor', 'tasa_euribor', 'num_empleados', 'nunca_contactado_antes']
Numéricas B: ['edad_cliente', 'num_contactos_campana_actual', 'num_contactos_campana_anterior', 'tasa_variacion_empleo', 'indice_precios_consumidor', 'indice_confianza_consumidor', 'tasa_euribor', 'num_empleados', 'nunca_contactado_antes']


In [58]:
X_A["nunca_contactado_antes"].value_counts()


nunca_contactado_antes
1    39661
0     1515
Name: count, dtype: int64

In [59]:
X_A["num_dias_desde_ultima_campana"].describe()


count    1515.000000
mean        6.014521
std         3.824906
min         0.000000
25%         3.000000
50%         6.000000
75%         7.000000
max        27.000000
Name: num_dias_desde_ultima_campana, dtype: float64

In [60]:
X_A["num_dias_desde_ultima_campana"].value_counts().sort_index()


num_dias_desde_ultima_campana
0.0      15
1.0      26
2.0      61
3.0     439
4.0     118
5.0      46
6.0     412
7.0      60
8.0      18
9.0      64
10.0     52
11.0     28
12.0     58
13.0     36
14.0     20
15.0     24
16.0     11
17.0      8
18.0      7
19.0      3
20.0      1
21.0      2
22.0      3
25.0      1
26.0      1
27.0      1
Name: count, dtype: int64

Imputación de nulos con la media

In [61]:
media_pdays = X_A["num_dias_desde_ultima_campana"].mean()

X_A["num_dias_desde_ultima_campana"] = (
    X_A["num_dias_desde_ultima_campana"]
    .fillna(media_pdays)
)


In [62]:
X_A["num_dias_desde_ultima_campana"].isna().sum()


np.int64(0)

In [63]:
X_A["num_dias_desde_ultima_campana"].describe()


count    41176.000000
mean         6.014521
std          0.733444
min          0.000000
25%          6.014521
50%          6.014521
75%          6.014521
max         27.000000
Name: num_dias_desde_ultima_campana, dtype: float64

Qué hemos hecho técnicamente

Eliminado el único NaN numérico.

Preparado df_A para algoritmos que no aceptan NaN.

Mantenido coherencia estadística.

Conservado la señal real con la variable binaria.

In [64]:
X_A[col_categoricas_A].isna().sum()


ocupacion                     0
estado_civil                  0
nivel_educacion               0
credito_actual                0
prestamo_hipotecario          0
prestamo_personal             0
tipo_contacto                 0
ultimo_mes_contacto           0
ultimo_dia_contacto           0
resultado_campana_anterior    0
dtype: int64

Transformar categóricas con One-Hot Encoding

In [65]:
X_A = pd.get_dummies(X_A, columns=col_categoricas_A, drop_first=True)


In [66]:
X_B = pd.get_dummies(X_B, columns=col_categoricas_B, drop_first=True)


In [67]:
X_A.head()



edad_cliente  num_contactos_campana_actual  num_dias_desde_ultima_campana  \
0            56                             1                       6.014521   
1            57                             1                       6.014521   
2            37                             1                       6.014521   
3            40                             1                       6.014521   
4            56                             1                       6.014521   

   num_contactos_campana_anterior  tasa_variacion_empleo  \
0                               0                    1.1   
1                               0                    1.1   
2                               0                    1.1   
3                               0                    1.1   
4                               0                    1.1   

   indice_precios_consumidor  indice_confianza_consumidor  tasa_euribor  \
0                     93.994                        -36.4         4.857   
1                     93.994                        -36.4         4.857   
2                     93.994                        -36.4         4.857   
3                     93.994                        -36.4         4.857   
4                     93.994                        -36.4         4.857   

   num_empleados  nunca_contactado_antes  ...  ultimo_mes_contacto_may  \
0         5191.0                       1  ...                     True   
1         5191.0                       1  ...                     True   
2         5191.0                       1  ...                     True   
3         5191.0                       1  ...                     True   
4         5191.0                       1  ...                     True   

   ultimo_mes_contacto_nov  ultimo_mes_contacto_oct  ultimo_mes_contacto_sep  \
0                    False                    False                    False   
1                    False                    False                    False   
2                    False                    False                    False   
3                    False                    False                    False   
4                    False                    False                    False   

   ultimo_dia_contacto_mon  ultimo_dia_contacto_thu  ultimo_dia_contacto_tue  \
0                     True                    False                    False   
1                     True                    False                    False   
2                     True                    False                    False   
3                     True                    False                    False   
4                     True                    False                    False   

   ultimo_dia_contacto_wed  resultado_campana_anterior_nonexistent  \
0                    False                                    True   
1                    False                                    True   
2                    False                                    True   
3                    False                                    True   
4                    False                                    True   

   resultado_campana_anterior_success  
0                               False  
1                               False  
2                               False  
3                               False  
4                               False  

[5 rows x 53 columns]

In [68]:
X_B.head()

edad_cliente  num_contactos_campana_actual  num_contactos_campana_anterior  \
0            56                             1                               0   
1            57                             1                               0   
2            37                             1                               0   
3            40                             1                               0   
4            56                             1                               0   

   tasa_variacion_empleo  indice_precios_consumidor  \
0                    1.1                     93.994   
1                    1.1                     93.994   
2                    1.1                     93.994   
3                    1.1                     93.994   
4                    1.1                     93.994   

   indice_confianza_consumidor  tasa_euribor  num_empleados  \
0                        -36.4         4.857         5191.0   
1                        -36.4         4.857         5191.0   
2                        -36.4         4.857         5191.0   
3                        -36.4         4.857         5191.0   
4                        -36.4         4.857         5191.0   

   nunca_contactado_antes  ocupacion_blue-collar  ...  \
0                       1                  False  ...   
1                       1                  False  ...   
2                       1                  False  ...   
3                       1                  False  ...   
4                       1                  False  ...   

   ultimo_mes_contacto_may  ultimo_mes_contacto_nov  ultimo_mes_contacto_oct  \
0                     True                    False                    False   
1                     True                    False                    False   
2                     True                    False                    False   
3                     True                    False                    False   
4                     True                    False                    False   

   ultimo_mes_contacto_sep  ultimo_dia_contacto_mon  ultimo_dia_contacto_thu  \
0                    False                     True                    False   
1                    False                     True                    False   
2                    False                     True                    False   
3                    False                     True                    False   
4                    False                     True                    False   

   ultimo_dia_contacto_tue  ultimo_dia_contacto_wed  \
0                    False                    False   
1                    False                    False   
2                    False                    False   
3                    False                    False   
4                    False                    False   

   resultado_campana_anterior_nonexistent  resultado_campana_anterior_success  
0                                    True                               False  
1                                    True                               False  
2                                    True                               False  
3                                    True                               False  
4                                    True                               False  

[5 rows x 52 columns]

In [69]:
X_A.select_dtypes(include=["object"]).columns


Index([], dtype='str')

In [70]:
X_A = X_A.astype(int)
X_B = X_B.astype(int)


In [71]:
X_A.dtypes.unique()


array([dtype('int64')], dtype=object)

In [72]:
X_B.dtypes.unique()

array([dtype('int64')], dtype=object)

# Conclusión del Preprocesamiento para Modelado

En esta fase se preparó el dataset para que pudiera ser utilizado correctamente por algoritmos de Machine Learning, asegurando que todas las variables cumplieran con los requisitos estructurales de los modelos.

Las acciones realizadas fueron:

* Eliminación de la variable duracion_contacto_segundos por data leakage, ya que representa información posterior a la decisión que el modelo pretende predecir.

* Transformación de num_dias_desde_ultima_campana, separando la condición “nunca contactado” en una variable binaria y tratando los valores faltantes de forma coherente.

* Creación de dos configuraciones de dataset:

    - df_A: mantiene la variable transformada.

    - df_B: elimina dicha variable para evaluar su impacto en el rendimiento del modelo.

* Separación formal de variables predictoras (X) y variable objetivo (y).

* Codificación de todas las variables categóricas mediante One-Hot Encoding.

* Conversión final del dataset en una matriz completamente numérica (int64), sin valores nulos ni variables de tipo texto.

Como resultado, se obtuvo un dataset estructuralmente limpio, consistente y apto para entrenamiento de modelos de clasificación, permitiendo comparar experimentalmente el impacto de ciertas decisiones de ingeniería de variables en el rendimiento predictivo.

# División Train / Test

In [73]:
from sklearn.model_selection import train_test_split

X_A_train, X_A_test, y_A_train, y_A_test = train_test_split(
    X_A,
    y_A,
    test_size=0.2,
    random_state=42,
    stratify=y_A
)

In [74]:
X_B_train, X_B_test, y_B_train, y_B_test = train_test_split(
    X_B,
    y_B,
    test_size=0.2,
    random_state=42,
    stratify=y_B
)


In [75]:
print("Distribución y_A_train:")
print(y_A_train.value_counts(normalize=True))

print("Distribución y_A_test:")
print(y_A_test.value_counts(normalize=True))


Distribución y_A_train:
contrata_deposito_largo_plazo
no     0.887341
yes    0.112659
Name: proportion, dtype: float64
Distribución y_A_test:
contrata_deposito_largo_plazo
no     0.887324
yes    0.112676
Name: proportion, dtype: float64


# Escalado

In [76]:
from sklearn.preprocessing import StandardScaler

scaler_A = StandardScaler()

# Ajustar SOLO con train
X_A_train_scaled = scaler_A.fit_transform(X_A_train)

# Transformar test usando parámetros del train
X_A_test_scaled = scaler_A.transform(X_A_test)


In [77]:
scaler_B = StandardScaler()

X_B_train_scaled = scaler_B.fit_transform(X_B_train)
X_B_test_scaled = scaler_B.transform(X_B_test)


In [78]:
X_A_train_scaled.mean(axis=0).mean()
X_A_train_scaled.std(axis=0).mean()


np.float64(1.0)

In [79]:
import pandas as pd

X_A_train_scaled = pd.DataFrame(X_A_train_scaled, columns=X_A_train.columns, index=X_A_train.index)
X_A_test_scaled = pd.DataFrame(X_A_test_scaled, columns=X_A_test.columns, index=X_A_test.index)

X_B_train_scaled = pd.DataFrame(X_B_train_scaled, columns=X_B_train.columns, index=X_B_train.index)
X_B_test_scaled = pd.DataFrame(X_B_test_scaled, columns=X_B_test.columns, index=X_B_test.index)


In [81]:
# =========================
# 0) IMPORTS (clasificación)
# =========================
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# ===========================================
# 1) SPLIT TRAIN -> (TRAIN + VALIDACIÓN)
#    (Usamos SOLO train para comparar modelos)
# ===========================================

# --- Modelo A ---
X_A_tr, X_A_val, y_A_tr, y_A_val = train_test_split(
    X_A_train_scaled, y_A_train,
    test_size=0.2, random_state=42, stratify=y_A_train
)

# --- Modelo B ---
X_B_tr, X_B_val, y_B_tr, y_B_val = train_test_split(
    X_B_train_scaled, y_B_train,
    test_size=0.2, random_state=42, stratify=y_B_train
)

# =========================
# 2) CREAR "CAJAS NEGRAS"
# =========================
knn_A = KNeighborsClassifier()
dt_A  = DecisionTreeClassifier(random_state=42)
lr_A  = LogisticRegression(max_iter=2000, random_state=42)
hgb_A = HistGradientBoostingClassifier(random_state=42)

knn_B = KNeighborsClassifier()
dt_B  = DecisionTreeClassifier(random_state=42)
lr_B  = LogisticRegression(max_iter=2000, random_state=42)
hgb_B = HistGradientBoostingClassifier(random_state=42)

# =========================
# 3) ENTRENAR
# =========================
knn_A.fit(X_A_tr, y_A_tr)
dt_A.fit(X_A_tr, y_A_tr)
lr_A.fit(X_A_tr, y_A_tr)
hgb_A.fit(X_A_tr, y_A_tr)

knn_B.fit(X_B_tr, y_B_tr)
dt_B.fit(X_B_tr, y_B_tr)
lr_B.fit(X_B_tr, y_B_tr)
hgb_B.fit(X_B_tr, y_B_tr)

# ===========================================
# 4) EVALUAR EN VALIDACIÓN Y SELECCIONAR MEJOR
# ===========================================

def evaluar_modelo(nombre, modelo, X_val, y_val):
    preds = modelo.predict(X_val)
    acc = accuracy_score(y_val, preds)
    cm = confusion_matrix(y_val, preds)
    acc_por_clase = cm.diagonal() / cm.sum(axis=1)
    print(f"\n{nombre}")
    print("Accuracy total:", acc)
    print("Accuracy por clase:", acc_por_clase)
    return acc, acc_por_clase, cm

print("========== RESULTADOS VALIDACIÓN - MODELO A ==========")
acc_knn_A, accpc_knn_A, cm_knn_A = evaluar_modelo("KNN (A)", knn_A, X_A_val, y_A_val)
acc_dt_A,  accpc_dt_A,  cm_dt_A  = evaluar_modelo("Árbol (A)", dt_A, X_A_val, y_A_val)
acc_lr_A,  accpc_lr_A,  cm_lr_A  = evaluar_modelo("Regresión Logística (A)", lr_A, X_A_val, y_A_val)
acc_hgb_A, accpc_hgb_A, cm_hgb_A = evaluar_modelo("HistGradientBoosting (A)", hgb_A, X_A_val, y_A_val)

print("\n========== RESULTADOS VALIDACIÓN - MODELO B ==========")
acc_knn_B, accpc_knn_B, cm_knn_B = evaluar_modelo("KNN (B)", knn_B, X_B_val, y_B_val)
acc_dt_B,  accpc_dt_B,  cm_dt_B  = evaluar_modelo("Árbol (B)", dt_B, X_B_val, y_B_val)
acc_lr_B,  accpc_lr_B,  cm_lr_B  = evaluar_modelo("Regresión Logística (B)", lr_B, X_B_val, y_B_val)
acc_hgb_B, accpc_hgb_B, cm_hgb_B = evaluar_modelo("HistGradientBoosting (B)", hgb_B, X_B_val, y_B_val)

# (Opcional) Selección automática por accuracy total
resultados_A = {
    "KNN": acc_knn_A,
    "DecisionTree": acc_dt_A,
    "LogisticRegression": acc_lr_A,
    "HistGradientBoosting": acc_hgb_A
}
mejor_A = max(resultados_A, key=resultados_A.get)
print("\nMejor modelo A por accuracy:", mejor_A, "=>", resultados_A[mejor_A])

resultados_B = {
    "KNN": acc_knn_B,
    "DecisionTree": acc_dt_B,
    "LogisticRegression": acc_lr_B,
    "HistGradientBoosting": acc_hgb_B
}
mejor_B = max(resultados_B, key=resultados_B.get)
print("Mejor modelo B por accuracy:", mejor_B, "=>", resultados_B[mejor_B])


========== RESULTADOS VALIDACIÓN - MODELO A ==========

KNN (A)
Accuracy total: 0.8940497874924105
Accuracy por clase: [0.97040712 0.29245283]

Árbol (A)
Accuracy total: 0.8386460230722526
Accuracy por clase: [0.90027369 0.35309973]

Regresión Logística (A)
Accuracy total: 0.9034608378870674
Accuracy por clase: [0.98905234 0.22911051]

HistGradientBoosting (A)
Accuracy total: 0.9020947176684881
Accuracy por clase: [0.9856312  0.24393531]

========== RESULTADOS VALIDACIÓN - MODELO B ==========

KNN (B)
Accuracy total: 0.8949605343047966
Accuracy por clase: [0.97160452 0.29110512]

Árbol (B)
Accuracy total: 0.8381906496660595
Accuracy por clase: [0.89856312 0.36253369]

Regresión Logística (B)
Accuracy total: 0.9033090467516697
Accuracy por clase: [0.98905234 0.2277628 ]

HistGradientBoosting (B)
Accuracy total: 0.9005768063145112
Accuracy por clase: [0.98426274 0.24123989]

Mejor modelo A por accuracy: LogisticRegression => 0.9034608378870674
Mejor modelo B por accuracy: LogisticRegress

# Resultados en test (modelo mostrado)

De tu salida:

🔹 Accuracy total:
0.9205

🔹 Accuracy por clase:
Clase 0 (no)  → 0.9659
Clase 1 (yes) → 0.5625

| Métrica   | Valor |
| --------- | ----- |
| Precision | 0.68  |
| Recall    | 0.56  |
| F1        | 0.61  |


# Interpretación importante

El dataset está desbalanceado:

88% → no

11% → yes

Por eso:

El accuracy total de 92% es engañoso.

Si el modelo predijera siempre “no”:

Tendría ≈ 88% accuracy.

Por tanto, lo que realmente importa aquí es:

Recall de la clase yes

F1-score de la clase yes

Porque el negocio quiere detectar clientes que sí contratan.

# ¿Cuál modelo seleccionar?

La decisión depende del objetivo del negocio.

🔹 Caso 1: Minimizar llamadas innecesarias

Queremos alta precisión en "yes".

→ Este modelo tiene precision 0.68
No está mal, pero tampoco excelente.

🔹 Caso 2: No perder clientes que sí contratarían

Queremos alto recall en "yes".

→ Recall = 0.56
Está dejando escapar casi la mitad.

🔹 Caso 3: Equilibrio general

F1 yes = 0.61
Es razonable, pero mejorable.

🧠 Interpretabilidad vs Rendimiento



KNN → interpretable parcialmente

Árbol → muy interpretable

Logística → muy interpretable

HistGradientBoosting → caja negra

En banca, muchas veces:

Se prefiere Regresión Logística por interpretabilidad y estabilidad regulatoria.

🎯 Conclusión profesional para este ejercicio

Si el modelo que dio 0.9205 fue HistGradientBoosting:

✔ Es el mejor en accuracy total
❗ Pero no necesariamente el mejor en recall de "yes"

Si Regresión Logística tiene resultados similares:

→ Sería preferible por interpretabilidad.

# Creación del predictor base (DummyClassifier = "moda") y evaluación en Dataset A y B

In [82]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


Baseline

In [83]:
# Baseline (moda) - Modelo A
dummy_A = DummyClassifier(strategy="most_frequent", random_state=42)
dummy_A.fit(X_A_train_scaled, y_A_train)

pred_A_base = dummy_A.predict(X_A_test_scaled)

acc_A_base = accuracy_score(y_A_test, pred_A_base)
cm_A_base = confusion_matrix(y_A_test, pred_A_base)

print("=== BASELINE Dummy (A) ===")
print("Accuracy base:", acc_A_base)
print("Accuracy por clase:", cm_A_base.diagonal() / cm_A_base.sum(axis=1))
print("\nClassification report:\n", classification_report(y_A_test, pred_A_base, zero_division=0))


=== BASELINE Dummy (A) ===
Accuracy base: 0.8873239436619719
Accuracy por clase: [1. 0.]

Classification report:
               precision    recall  f1-score   support

          no       0.89      1.00      0.94      7308
         yes       0.00      0.00      0.00       928

    accuracy                           0.89      8236
   macro avg       0.44      0.50      0.47      8236
weighted avg       0.79      0.89      0.83      8236



In [84]:
# Baseline (moda) - Modelo B
dummy_B = DummyClassifier(strategy="most_frequent", random_state=42)
dummy_B.fit(X_B_train_scaled, y_B_train)

pred_B_base = dummy_B.predict(X_B_test_scaled)

acc_B_base = accuracy_score(y_B_test, pred_B_base)
cm_B_base = confusion_matrix(y_B_test, pred_B_base)

print("=== BASELINE Dummy (B) ===")
print("Accuracy base:", acc_B_base)
print("Accuracy por clase:", cm_B_base.diagonal() / cm_B_base.sum(axis=1))
print("\nClassification report:\n", classification_report(y_B_test, pred_B_base, zero_division=0))


=== BASELINE Dummy (B) ===
Accuracy base: 0.8873239436619719
Accuracy por clase: [1. 0.]

Classification report:
               precision    recall  f1-score   support

          no       0.89      1.00      0.94      7308
         yes       0.00      0.00      0.00       928

    accuracy                           0.89      8236
   macro avg       0.44      0.50      0.47      8236
weighted avg       0.79      0.89      0.83      8236



# Interpretación

El modelo base:

Predice siempre "no"

Acierta todos los "no"

No detecta ningún "yes"

Tiene 88.7% accuracy solo por desbalance

Este número (0.887) es el mínimo aceptable.

Cualquier modelo que:

No supere 0.887 → no está aprendiendo.

No mejore recall de "yes" → no aporta valor.

Comparación con tu modelo real

Tu modelo real:

Accuracy = 0.9205

Recall yes = 0.56

F1 yes = 0.61

Comparado con baseline

| Métrica    | Baseline | Modelo |
| ---------- | -------- | ------ |
| Accuracy   | 0.887    | 0.920  |
| Recall yes | 0.00     | 0.56   |
| F1 yes     | 0.00     | 0.61   |


✔ Mejora real

✔ Detecta clientes que contratarán

✔ Genera valor de negocio

# Conclusión

El modelo predictivo supera ampliamente al baseline basado en la clase mayoritaria. Mientras que el DummyClassifier no identifica ningún cliente que contrate el depósito (recall=0), el modelo entrenado logra detectar el 56% de ellos, demostrando que realmente aprende patrones relevantes y aporta valor frente a una estrategia trivial.